plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
+        return key;
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static AwsCredentialsProvider createAwsCredentialsProvider(ExchangeS3Config config)
+    {
+        if (config.getS3AwsAccessKey() != null && config.getS3AwsSecretKey() != null) {
+            return StaticCredentialsProvider.create(AwsBasicCredentials.create(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        return DefaultCredentialsProvider.create();
+    }
+
+    private S3Client createS3Client(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3Client s3Client;
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private int currentPartNumber;
+        private Optional<String> uploadId = Optional.empty();
+        private final List<CompletableFuture<CompletedPart>> uploadFutures = new ArrayList<>();
+
+        public S3ExchangeStorageWriter(S3Client s3Client, S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3Client = requireNonNull(s3Client, "s3Client is null");
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
+                throws IOException
```

Comment:
I wonder if this interface has to throw an exception directly? Or should it return a failed future in case of an exception? Usually the second one is preferred.

Reply:
Sure, we can do that easily